### PR TITLE
Request To Merge With The Master Branch

### PR DESCRIPTION
**Pull request by mohamednaser**
I have developed the below tasks :

[x] Create my_each Function Click For More Info
[x] Create my_each_with_index function
[x] Create my_select function Click For More Info
[x] Create my_all function Click For More Info
[x] Create my_any function Click For More Info
[x] Create my_none function Click For More Info
[x] Create my_count function Click For More Info
[x] Create my_map function Click For More Info
[x] Create my_inject function Click For More Info
[x] Modify my_map function to accept block or proc parameter Click For More Info

and waiting for your feedback if you have any comments or any feedback to improve this code base

### CODE REVIEW

File: index.rb [hunk 1]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+      i += 1
+    end
+    array2
+  end
+
+  def my_inject
+    result = 0
+    my_each { |element| result = yield(result, element) }
+    result
+  end
```
Comment (maelfosso) [inline index.rb]:
```ruby
irb(main):129:0> range = Range.new(5, 50)
=> 5..50
irb(main):130:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
irb(main):131:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):131
        1: from index.rb:63:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):132:0> 
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):134:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):135:0> array.inject(:+)
=> 377
irb(main):136:0> array.my_inject(:+)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):136
        1: from index.rb:63:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):137:0> 
```

File: index.rb [hunk 2]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+    while i < size
+      array2 << if block_given?
+                  yield(self[i])
+                else
+                  proc.call(self[i])
+                end
+      i += 1
+    end
+    array2
+  end
```
Comment (maelfosso) [inline index.rb]: Should return an **enumerator**
```ruby
irb(main):123:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):124:0> array.map
=> #<Enumerator: [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]:map>
irb(main):125:0> array.my_map
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):125
        2: from index.rb:56:in `my_map'
        1: from index.rb:56:in `proc'
ArgumentError (tried to create Proc object without a block)
irb(main):126:0> 
```

File: index.rb [hunk 3]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+
+  def my_none
+    my_each { |element| return false if yield(element) == true }
+    true
+  end
+
+  def my_count
+    size
+  end
```
Comment (maelfosso) [inline index.rb]:
```ruby
irb(main):113:0> array.count(0)
=> 9
irb(main):114:0> array.my_count(0)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):114
        1: from index.rb:45:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):115:0> 
```
You have to count the number of element equal to **0**

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):117:0> block = proc { |num| num < 4 }
=> #<Proc:0x00005619b0c31250@(irb):117>
irb(main):118:0> array.count(&block)
=> 47
irb(main):119:0> array.my_count(&block)
=> 100
```
You have to count the number of element that satisfy the condition ```num < 4```

File: index.rb [hunk 4]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+
+  def my_any
+    my_each { |element| return true unless yield(element) == false }
+    false
+  end
+
+  def my_none
+    my_each { |element| return false if yield(element) == true }
+    true
+  end
```
Comment (maelfosso) [inline index.rb]:
```ruby
irb(main):106:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):107:0> words[0] = 5
=> 5
irb(main):108:0> words.none?(5)
=> false
irb(main):109:0> words.my_none(5)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):109
        1: from index.rb:40:in `my_none'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):110:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):101:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):102:0> words.none?(/z/)
=> true
irb(main):103:0> words.my_none(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):103
        1: from index.rb:40:in `my_none'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):104:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):095:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):096:0> array.none?(String)
=> true
irb(main):097:0> array.my_none(String)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):97
        1: from index.rb:40:in `my_none'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):098:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):090:0> false_array = [nil, false, nil, false]
=> [nil, false, nil, false]
irb(main):091:0> false_array.none?
=> true
irb(main):092:0> false_array.my_none
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):92
        3: from index.rb:41:in `my_none'
        2: from index.rb:8:in `my_each'
        1: from index.rb:41:in `block in my_none'
LocalJumpError (no block given (yield))
irb(main):093:0>
```

File: index.rb [hunk 5]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+  def my_all
+    my_each { |element| return false unless yield(element) == true }
+
+    true
+  end
+
+  def my_any
+    my_each { |element| return true unless yield(element) == false }
+    false
+  end
```
Comment (maelfosso) [inline index.rb]:
```ruby
irb(main):083:0> false_array = [nil, false, nil, false]
=> [nil, false, nil, false]
irb(main):084:0> false_array.any?
=> false
irb(main):085:0> false_array.my_any
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):85
        3: from index.rb:36:in `my_any'
        2: from index.rb:8:in `my_each'
        1: from index.rb:36:in `block in my_any'
LocalJumpError (no block given (yield))
irb(main):086:0> 
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):071:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):072:0> words.any?(/z/)
=> false
irb(main):073:0> words.my_any(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):73
        1: from index.rb:35:in `my_any'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):074:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):076:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):077:0> words[0] = 'cat'
=> "cat"
irb(main):078:0> words.any?('cat')
=> true
irb(main):079:0> words.my_any('cat')
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):79
        1: from index.rb:35:in `my_any'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):080:0> 
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):061:0> true_array = [nil, false, true, []]
=> [nil, false, true, []]
irb(main):062:0> true_array.any?
=> true
irb(main):063:0> true_array.my_any
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):63
        3: from index.rb:36:in `my_any'
        2: from index.rb:8:in `my_each'
        1: from index.rb:36:in `block in my_any'
LocalJumpError (no block given (yield))
irb(main):064:0> 
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):066:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):067:0> array.any?(Integer)
=> true
irb(main):068:0> array.my_any(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):68
        1: from index.rb:35:in `my_any'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):069:0> 
```

File: index.rb [hunk 6]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+    my_each { |element| new_array << element if yield(element) }
+
+    new_array
+  end
+
+  def my_all
+    my_each { |element| return false unless yield(element) == true }
+
+    true
+  end
```
Comment (maelfosso) [inline index.rb]:
```ruby
irb(main):038:0> true_array = [nil, false, true, []]
=> [nil, false, true, []]
irb(main):039:0> true_array.all?
=> false
irb(main):040:0> true_array.my_all
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):40
        3: from index.rb:30:in `my_all'
        2: from index.rb:8:in `my_each'
        1: from index.rb:30:in `block in my_all'
LocalJumpError (no block given (yield))
irb(main):041:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):043:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):044:0> array.all?(Integer)
=> true
irb(main):045:0> array.my_all(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):45
        1: from index.rb:29:in `my_all'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):046:0>
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):048:0> words.all?(/d/)
=> true
irb(main):049:0> words.my_all(/d/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):49
        1: from index.rb:29:in `my_all'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):050:0> 
```

Comment (maelfosso) [reply index.rb]:
```ruby
irb(main):055:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):056:0> array.all?(3)
=> false
irb(main):057:0> array.my_all(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):57
        1: from index.rb:29:in `my_all'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):058:0>
```

File: index.rb [hunk 7]
```diff
@@ -9,10 +9,113 @@ def my_each
...
+    end
+  end
+
+  def my_select
+    new_array = []
+
+    my_each { |element| new_array << element if yield(element) }
+
+    new_array
+  end
```
Comment (maelfosso) [inline index.rb]: Should return an **enumerator**
```ruby
irb(main):023:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):024:0> array.select
=> #<Enumerator: [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]:select>
irb(main):025:0> array.my_select
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):25
        3: from index.rb:24:in `my_select'
        2: from index.rb:8:in `my_each'
        1: from index.rb:24:in `block in my_select'
LocalJumpError (no block given (yield))
irb(main):026:0>
```

File: index.rb [hunk 8]
```diff
@@ -9,10 +9,113 @@ def my_each
...
     end
   end
+
+  def my_each_with_index
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+  end
```
Comment (maelfosso) [inline index.rb]: Should return an enumerator
```ruby
irb(main):018:0> array.each_with_index
=> #<Enumerator: [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]:each_with_index>
irb(main):019:0> array.my_each_with_index
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):19
        1: from index.rb:16:in `my_each_with_index'
LocalJumpError (no block given (yield))
irb(main):020:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):021:0>
```

File: index.rb [hunk 9]
```diff
@@ -9,10 +9,113 @@ def my_each
       i += 1
     end
   end
```
Comment (maelfosso) [inline index.rb]: Should return an enumerator
```ruby
irb(main):013:0> array
=> [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]
irb(main):014:0> array.each
=> #<Enumerator: [1, 2, 1, 8, 4, 5, 3, 0, 0, 2, 8, 7, 8, 6, 2, 6, 2, 7, 7, 4, 3, 0, 3, 2, 2, 2, 0, 1, 6, 8, 5, 4, 5, 4, 3, 5, 6, 4, 2, 3, 5, 2, 6, 1, 5, 4, 0, 6, 6, 5, 8, 5, 2, 1, 2, 8, 0, 6, 3, 1, 7, 5, 2, 4, 0, 5, 5, 6, 4, 2, 3, 3, 1, 3, 2, 6, 2, 8, 3, 0, 6, 3, 4, 1, 4, 1, 1, 6, 4, 5, 3, 6, 5, 0, 5, 7, 5, 2, 7, 4]:each>
irb(main):015:0> array.my_each
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):15
        1: from index.rb:8:in `my_each'
LocalJumpError (no block given (yield))
irb(main):016:0> 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @mohamednaser 

Nice introduction to the project :clap: 
However there are some of bug to fix. I invite you to read the comments to learn more about those bugs.

Kindly make the changes and submit a new code review request.
Thanks,

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: libs/enumerable_conditions.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    end
+
+    if param.nil?
+      my_each { |element| return true if element }
+      return false
+    end
+
+    my_each { |element| return true if check_input_valid_by_critira(element, param) }
+    false
+  end
```
Comment (maelfosso) [inline libs/enumerable_conditions.rb]:
```ruby
irb(main):008:0> false_block = proc { |num| num > 9 }
=> #<Proc:0x00005566f019a9b8@(irb):8>
irb(main):009:0> array
=> [2, 0, 8, 6, 1, 1, 2, 5, 1, 3, 7, 6, 1, 3, 1, 6, 0, 2, 2, 7, 7, 8, 8, 6, 2, 0, 1, 6, 1, 2, 1, 7, 7, 0, 7, 6, 5, 3, 1, 0, 7, 7, 4, 7, 4, 1, 3, 3, 0, 4, 4, 2, 8, 6, 3, 5, 4, 4, 0, 7, 0, 8, 5, 6, 3, 7, 6, 2, 2, 1, 1, 8, 7, 2, 3, 3, 8, 6, 7, 6, 2, 0, 5, 7, 7, 8, 3, 3, 7, 8, 1, 6, 8, 0, 5, 6, 8, 4, 6, 6]
irb(main):010:0> array.any?(&false_block)
=> false
irb(main):011:0> array.my_any?(&false_block)
Traceback (most recent call last):
        2: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        1: from (irb):11
NoMethodError (undefined method `my_any?' for #<Array:0x00005566f01ed348>)
Did you mean?  my_any
irb(main):012:0> array.my_any(&false_block)
=> true
irb(main):013:0>
```

File: libs/enumerable_conditions.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    end
+
+    if param.nil?
+      my_each { |element| return false unless element }
+      return true
+    end
+
+    my_each { |element| return false if check_input_valid_by_critira(element, param) }
+    true
+  end
```
Comment (maelfosso) [inline libs/enumerable_conditions.rb]:
```ruby
irb(main):015:0> false_array = [nil, false, nil, false]
=> [nil, false, nil, false]
irb(main):016:0> false_array.none?
=> true
irb(main):017:0> false_array.my_none
=> false
irb(main):018:0>
```

File: libs/enumerable_helpers.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      end
+
+      return result
+    end
+
+    return if param.is_a? Symbol
+
+    array_input[1..-1].my_each do |element|
+      result = result.send(param, element)
+    end
```
Comment (maelfosso) [inline libs/enumerable_helpers.rb]:
```ruby
irb(main):021:0> array
=> [2, 0, 8, 6, 1, 1, 2, 5, 1, 3, 7, 6, 1, 3, 1, 6, 0, 2, 2, 7, 7, 8, 8, 6, 2, 0, 1, 6, 1, 2, 1, 7, 7, 0, 7, 6, 5, 3, 1, 0, 7, 7, 4, 7, 4, 1, 3, 3, 0, 4, 4, 2, 8, 6, 3, 5, 4, 4, 0, 7, 0, 8, 5, 6, 3, 7, 6, 2, 2, 1, 1, 8, 7, 2, 3, 3, 8, 6, 7, 6, 2, 0, 5, 7, 7, 8, 3, 3, 7, 8, 1, 6, 8, 0, 5, 6, 8, 4, 6, 6]
irb(main):022:0> array.inject(:+)
=> 416
irb(main):023:0> array.my_inject(:+)
=> nil
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @mohamednaser 

Please, receive all my congratulations for that huge jump you did forward :handshake: 
However, there are still some bugs to fix, exactly 3. You can read more about it in the comments left below.
Kindly make the changes and submit a new code review requests.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice :+1: 
you have implemented most requirements. That's good.
However, I found one bug with ```my_none``` method. Have a look at the screenshot.
Screenshot from 2019-11-06 22-08-55

#### Fix that and re-submit.

### Martin Afani

File: libs/enumerable_conditions.rb [hunk 13]
```diff
@@ -30,8 +30,8 @@ def my_any(param = true, &proc)
 
   def my_none(param = nil)
     if block_given?
-      my_each { |element| return false if yield(element) == true }
-      true
+      my_each { |element| return false if yield(element)}
```
Comment (samgaco) [inline libs/enumerable_conditions.rb]: :+1:

**Review by samgaco — APPROVED** [review]
Everything looks fine now :+1:  You can move on.

-----------------------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:

**Pull request merged**